geoips/commandline/geoips_command.py

```diff
@@ -0,0 +1,495 @@
+"""
+
+import abc
+import argparse
+from colorama import Fore, Style
+from importlib import resources
+import json
+from os.path import dirname
+from shutil import get_terminal_size
+from tabulate import tabulate
+import yaml
+
+from geoips.commandline.ancillary_info import cmd_instructions
+from geoips.geoips_utils import get_entry_point_group
+
+
+class PluginPackages:
+    """Class to hold the plugin packages and their paths.
+
+    This class is used to hold the plugin packages and their paths, which are used
+    throughout the CLI to determine which plugins are available to the user. This
+    class is used in the GeoipsCLI class to determine which plugins are available to
+    the user.
+
+    Moving this to a class reduced GeoIPS CLI startup time by about 30%.
+    """
+
+    def __init__(self):
+        """Initialize the PluginPackages class.
+
+        Initialize the plugin packages and their paths. This is done by using the
+        get_plugin_packages() and get_plugin_package_paths() functions.
+        """
+        self.entrypoints = [
+            ep.value for ep in get_entry_point_group("geoips.plugin_packages")
+        ]
+        self.paths = [
+            dirname(resources.files(ep.value))
+            for ep in get_entry_point_group("geoips.plugin_packages")
+        ]
+
+
+plugin_packages = PluginPackages()
+
+
+class GeoipsCommand(abc.ABC):
+    """Abstract Base Class for top-level GeoIPS Command Classes, such as get or list.
+
+    This class is a blueprint of what each top-level GeoIPS Command Classes should
+    implement. Includes shared attributes and an ``add_suparsers`` function which is
+    used for initializing sub-command classes of a certain GeoIPS Command.
+    """
+
+    def __init__(self, parent=None, legacy=False):
+        """Initialize GeoipsCommand with a subparser and default to the command func.
+
+        Do this for each GeoipsCLI.geoips_subcommand_classes. This will instantiate
+        each subcommand class with a parser and point towards the correct default
+        function to call if that subcommand has been called.
+        """
+        self.legacy = legacy
+        self.nrl_url = "https://github.com/NRLMMD-GEOIPS/"
+        self.parent = parent
+        if self.parent:
+            if self.parent.subcommand_name == "cli":
+                combined_name = self.subcommand_name
+                parent_parsers = []
+            else:
+                combined_name = f"{self.parent.subcommand_name}_{self.subcommand_name}"
+                # We need to create a Geoips<cmd>Common Class for arguments
+                # that are shared between common commands. For example, we've created
+                # a 'GeoipsListCommon' class which adds arguments that will be shared
+                # by each GeoipsList<sub-cmd> class. Ie. if GeoipsListCommon has
+                # arguments --package, --columns, etc., and all of those arguments
+                # would be inherited by each GeoipsList<sub-cmd>
```

Comment:
Is there an issue for this?

Reply:
Can you create an issue to just discuss how to implement options for different scopes? I think we can make use of `ArgumentParser.parse_known_args` to help implement this.

Thinking about it, maybe there is already an issue?

Reply:
I say more about this here: https://github.com/NRLMMD-GEOIPS/geoips/pull/465#discussion_r1619204653

Reply:
See #637.